### PR TITLE
added ShoppingCart

### PR DESCRIPTION
**Pull request by chertterner**

### CODE REVIEW

File: src/main/java/com/example/bookshop/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    @GetMapping
+    public Page<ShoppingCartDto> getAll(Pageable pageable) {
+        return shoppingCartService.findAll(pageable);
+    }
+
+    @Operation(summary = "Add book to the shopping cart",
+            description = "Add new book to the shopping cart")
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping
+    public void addBookShoppingCart(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto addBookShoppingCart(
```

File: src/main/java/com/example/bookshop/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            @RequestBody @Valid AddBookShoppingCartRequestDto
+                    addBookShoppingCartRequestDto,
+            @AuthenticationPrincipal UserDetails userDetails) {
+        shoppingCartService.save(addBookShoppingCartRequestDto, userDetails);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @Operation(summary = "Update quantity of a book", description = "Update quantity of a book")
+    @PutMapping("/items/{id}")
+    public void updateCategory(@PathVariable Long id,
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/controller/ShoppingCartController.java]: rename method
```suggestion
    public ShoppingCartDto updateCategory(@PathVariable Long id,
```

File: src/main/java/com/example/bookshop/dto/AddBookShoppingCartRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.example.bookshop.dto;
+
+import lombok.Data;
+
+@Data
+public class AddBookShoppingCartRequestDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/dto/AddBookShoppingCartRequestDto.java]:
```suggestion
public class CartItemRequestDto {
```

File: src/main/java/com/example/bookshop/dto/AddBookShoppingCartRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+    private long bookId;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/dto/AddBookShoppingCartRequestDto.java]:
```suggestion
    private Long bookId;
```

Comment (Elena-Bruyako) [reply src/main/java/com/example/bookshop/dto/AddBookShoppingCartRequestDto.java]: add validation for bookId and quantity

File: src/main/java/com/example/bookshop/dto/BookQuantityReqDto.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.example.bookshop.dto;
+
+import lombok.Data;
+
+@Data
+public class BookQuantityReqDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/dto/BookQuantityReqDto.java]:
```suggestion
public class UpdateCartItemDto {
```

File: src/main/java/com/example/bookshop/dto/BookQuantityReqDto.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+    private int quantity;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/dto/BookQuantityReqDto.java]: add validation

File: src/main/java/com/example/bookshop/model/Book.java [hunk 7]
```diff
@@ -51,4 +52,7 @@ public class Book {
     @ToString.Exclude
     @EqualsAndHashCode.Exclude
     private Set<Category> categories = new HashSet<>();
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/Book.java]:
```suggestion
    @OneToOne(fetch = FetchType.LAZY)
```

File: src/main/java/com/example/bookshop/model/CartItem.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+@Table(name = "cart_items")
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id = ?")
+@SQLRestriction("status <> 'DELETED'")
+@Getter
+@Setter
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/example/bookshop/model/CartItem.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    @JoinColumn(name = "shopping_cart", nullable = false)
+    private ShoppingCart shoppingCart;
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/CartItem.java]: add Lazy and JoinColumn

File: src/main/java/com/example/bookshop/model/CartItem.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    private Book book;
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/CartItem.java]:
```suggestion
```

File: src/main/java/com/example/bookshop/model/ShoppingCart.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+@Setter
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @NotNull
+    @OneToOne
+    private User user;
+    @OneToMany
+    private Set<CartItem> cartItems;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/ShoppingCart.java]:
```suggestion
    private Set<CartItem> cartItems = new HashSet<>();
```

File: src/main/java/com/example/bookshop/model/ShoppingCart.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+@Table(name = "shopping_cart")
+@SQLDelete(sql = "UPDATE shopping_cart SET  is_deleted = true WHERE id = ?")
+@SQLRestriction("status <> 'DELETED'")
+@Getter
+@Setter
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/ShoppingCart.java]:
```suggestion
```

File: src/main/java/com/example/bookshop/model/ShoppingCart.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/ShoppingCart.java]: you can use @MapsId here 
```suggestion
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id", nullable = false)
```

File: src/main/java/com/example/bookshop/model/User.java [hunk 14]
```diff
@@ -40,17 +41,18 @@ public class User implements UserDetails {
...
     @ManyToMany
     @JoinTable(
             name = "users_roles",
             joinColumns = @JoinColumn(name = "user_id"),
             inverseJoinColumns = @JoinColumn(name = "role_id")
     )
     private Set<Role> roles = new HashSet<>();
+    @OneToOne
+    @JoinColumn(name = "shopping_cart", referencedColumnName = "id")
+    private ShoppingCart shoppingCart;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/User.java]: remove

File: src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+import com.example.bookshop.service.ShoppingCartService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java]: Think where `@Transactional` is needed

File: src/main/java/com/example/bookshop/model/ShoppingCart.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    private User user;
+    @OneToMany
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+                .stream()
+                .map(shoppingCartMapper::toDto)
+                .toList());
+    }
+
+    @Override
+    public void save(AddBookShoppingCartRequestDto addBookShoppingCartRequestDto,
+                                UserDetails userDetails) {
+        Book book = bookRepository.getReferenceById(addBookShoppingCartRequestDto.getBookId());
+        CartItem cartItem = new CartItem();
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java]: find shop cart by id, check if cartItems in shop cart already include this book,  add book, save shop cart to db

File: src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+        cartItem.setBook(book);
+        cartItem.setQuantity(addBookShoppingCartRequestDto.getQuantity());
+        User user = getCurrentUser(userDetails);
+        user.getShoppingCart().getCartItems().add(cartItem);
+        userRepository.save(user);
+    }
+
+    @Override
+    public void updateBookQuantity(Long id, int quantity, UserDetails userDetails) {
+        User user = getCurrentUser(userDetails);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java]: let's find shop cart by userId
find cartitem `findByIdAndShoppingCartId`
update quantity 
save cartItem

File: src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+                .getCartItems()
+                .stream()
+                .filter(cartItem -> cartItem.getId().equals(id))
+                .forEach(cartItem -> cartItem.setQuantity(quantity));
+        userRepository.save(user);
+    }
+
+    @Override
+    public void deleteBook(Long id, UserDetails userDetails) {
+        User user = getCurrentUser(userDetails);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java]: same here
find shop cart by userId
find cart item `findByIdAndShoppingCartId`
remove

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/example/bookshop/dto/CartItemRequestDto.java [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+import org.hibernate.validator.constraints.Length;
+
+@Data
+public class CartItemRequestDto {
+    @NotNull
+    private Long bookId;
+    @Length(min = 1)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/dto/CartItemRequestDto.java]:
```suggestion
    @NotNull
    @Positive
    private Long bookId;
    @Positive
```

File: src/main/java/com/example/bookshop/dto/UpdateCartItemDto.java [hunk 21]
```diff
@@ -0,0 +1,10 @@
+package com.example.bookshop.dto;
+
+import lombok.Data;
+import org.hibernate.validator.constraints.Length;
+
+@Data
+public class UpdateCartItemDto {
+    @Length(min = 1)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/dto/UpdateCartItemDto.java]:
```suggestion
    @Positive
```

File: src/main/java/com/example/bookshop/model/Book.java [hunk 22]
```diff
@@ -51,4 +53,7 @@ public class Book {
     @ToString.Exclude
     @EqualsAndHashCode.Exclude
     private Set<Category> categories = new HashSet<>();
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "cart_item", referencedColumnName = "id")
+    private CartItem cartItem;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/Book.java]:
```suggestion
    @OneToMany(mappedBy = "book", cascade = CascadeType.REMOVE, orphanRemoval = true)
    private List<CartItem> cartItems = new ArrayList<>();
```

File: src/main/java/com/example/bookshop/model/ShoppingCart.java [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+import jakarta.persistence.Table;
+import java.util.HashSet;
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Table(name = "shopping_cart")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/ShoppingCart.java]: table name should be plural
```suggestion
@Table(name = "shopping_carts")
```

File: src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+@Transactional
+public class ShoppingCartImpl implements ShoppingCartService {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java]:
```suggestion
public class ShoppingCartServiceImpl implements ShoppingCartService {
```

File: src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+    public void deleteBook(Long id, UserDetails userDetails) {
+        User user = getCurrentUser(userDetails);
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUser_Id(user.getId());
+        CartItem cartItem = cartItemRepository.findByBook_IdAndShoppingCart_Id(
+                id, shoppingCart.getId()
+        );
+        cartItemRepository.delete(cartItem);
+    }
+
+    private User getCurrentUser(UserDetails userDetails) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/service/implementation/ShoppingCartImpl.java]: you can provide Authentication to endpoints and get User from Principal and find user by id

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/example/bookshop/dto/cartitem/CartItemDto.java [hunk 26]
```diff
@@ -0,0 +1,19 @@
+package com.example.bookshop.dto.cartitem;
+
+import com.example.bookshop.model.Book;
+import com.example.bookshop.model.ShoppingCart;
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+import lombok.NonNull;
+
+@Data
+public class CartItemDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/dto/cartitem/CartItemDto.java]: Do you use this DTO?

File: src/main/java/com/example/bookshop/dto/cartitem/ShoppingCartDto.java [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+
+import com.example.bookshop.model.CartItem;
+import com.example.bookshop.model.User;
+import jakarta.validation.constraints.NotNull;
+import java.util.Set;
+import lombok.Data;
+
+@Data
+public class ShoppingCartDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/dto/cartitem/ShoppingCartDto.java]: don't need to validate response

File: src/main/java/com/example/bookshop/model/ShoppingCart.java [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+@Table(name = "shopping_carts")
+@SQLDelete(sql = "UPDATE shopping_carts SET  is_deleted = true WHERE id = ?")
+@SQLRestriction("status <> 'DELETED'")
+@Getter
+@Setter
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/ShoppingCart.java:27]: you can use `@MapsId` here

File: src/main/java/com/example/bookshop/model/ShoppingCart.java [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+    @JoinColumn(name = "id", nullable = false)
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
+    private Set<CartItem> cartItems = new HashSet<>();
+}
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/model/ShoppingCart.java:33]: do we need soft delete approach here?

File: src/main/java/com/example/bookshop/service/implementation/ShoppingCartServiceImpl.java [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+                .toList());
+    }
+
+    @Override
+    public ShoppingCartDto save(CartItemRequestDto cartItemRequestDto,
+                                User user) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUser_Id(
+                user.getId()
+        );
+        Optional<CartItem> cartItem = Optional.ofNullable(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/service/implementation/ShoppingCartServiceImpl.java]: just `shoppingCart.getCartItems()`

File: src/main/java/com/example/bookshop/service/implementation/ShoppingCartServiceImpl.java [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+@Service
+@RequiredArgsConstructor
+@Transactional
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private ShoppingCartRepository shoppingCartRepository;
+    private ShoppingCartMapper shoppingCartMapper;
+    private BookRepository bookRepository;
+    private UserRepository userRepository;
+    private CartItemRepository cartItemRepository;
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/service/implementation/ShoppingCartServiceImpl.java:32]: let's create shoppingCart when user registers. Create method in `ShoppingCartService` and use it `UserService`

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**